toplevel/genprintval.ml

```diff
@@ -613,6 +613,9 @@ module Make(O : OBJ)(EVP : EVALPATH with type valu = O.t) = struct
             Oval_printer printer)
 
 
-    in nest tree_of_val max_depth obj ty
-
+    in
+    let snap = Btype.snapshot () in
+    Misc.try_finally
+      (fun () -> nest tree_of_val max_depth obj (Ctype.correct_levels ty))
+      ~always:(fun () -> Btype.backtrack snap)
```

Comment:
This is presumably an unrelated fix that you needed for some expect tests to work correctly?

Reply:
Doing nothing changes the output in `typing-gadts/principality-and-gadts.ml`, suggesting that the printing function alters the type in an observable way.
However, if we use `Ctype.correct_levels`, there doesn't seem to be a need to backtrack (the above examples does not change), so I simplified a bit the code.